src/zon.zig

```diff
@@ -0,0 +1,831 @@
+    const ip = &self.sema.pt.zcu.intern_pool;
+    const gpa = self.sema.gpa;
+
+    const ptr_info = res_ty.ptrInfo(self.sema.pt.zcu);
+
+    if (ptr_info.flags.size != .Slice) {
+        return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "non slice pointers are not available in ZON",
+            .{},
+        );
+    }
+
+    // String literals
+    const string_alignment = ptr_info.flags.alignment == .none or ptr_info.flags.alignment == .@"1";
+    const string_sentinel = ptr_info.sentinel == .none or ptr_info.sentinel == .zero_u8;
+    if (string_alignment and ptr_info.child == .u8_type and string_sentinel) {
+        switch (node.get(self.file.zoir.?)) {
+            .string_literal => |val| {
+                const ip_str = try ip.getOrPutString(gpa, self.sema.pt.tid, val, .maybe_embedded_nulls);
+                const str_ref = try self.sema.addStrLit(ip_str, val.len);
+                return (try self.sema.coerce(
+                    self.block,
+                    res_ty,
+                    str_ref,
+                    try self.lazySrcLoc(.{ .node_abs = node.getAstNode(self.file.zoir.?) }),
+                )).toInterned().?;
+            },
+            else => {},
+        }
+    }
+
+    // Slice literals
+    const elem_nodes: Zoir.Node.Index.Range = switch (node.get(self.file.zoir.?)) {
+        .array_literal => |nodes| nodes,
+        .empty_literal => .{ .start = node, .len = 0 },
+        else => return self.fail(
+            .{ .node_abs = node.getAstNode(self.file.zoir.?) },
+            "expected type '{}'",
+            .{res_ty.fmt(self.sema.pt)},
+        ),
+    };
+
+    const elems = try self.sema.arena.alloc(InternPool.Index, elem_nodes.len + @intFromBool(ptr_info.sentinel != .none));
+
+    for (0..elem_nodes.len) |i| {
+        elems[i] = try self.lowerExpr(elem_nodes.at(@intCast(i)), Type.fromInterned(ptr_info.child));
+    }
+
+    if (ptr_info.sentinel != .none) {
+        elems[elems.len - 1] = ptr_info.sentinel;
+    }
+
+    const array_ty = try self.sema.pt.intern(.{ .array_type = .{
+        .len = elems.len,
+        .sentinel = ptr_info.sentinel,
+        .child = ptr_info.child,
+    } });
+
+    const array = try self.sema.pt.intern(.{ .aggregate = .{
+        .ty = array_ty,
+        .storage = .{ .elems = elems },
+    } });
+
+    const many_item_ptr_type = try ip.get(gpa, self.sema.pt.tid, .{ .ptr_type = .{
+        .child = ptr_info.child,
+        .sentinel = ptr_info.sentinel,
+        .flags = b: {
+            var flags = ptr_info.flags;
+            flags.size = .Many;
+            break :b flags;
+        },
+        .packed_offset = ptr_info.packed_offset,
+    } });
+
+    const many_item_ptr = try ip.get(gpa, self.sema.pt.tid, .{
+        .ptr = .{
+            .ty = many_item_ptr_type,
+            .base_addr = .{
+                .uav = .{
+                    .orig_ty = res_ty.toIntern(),
```

Comment:
This definitely isn't correct, I'm surprised it's not tripping any assertions. I think this should be `pt.singleConstPtrType(array_ty)` -- see `Zcu.PerThread.refValue` for an example.

Reply:
Could you elaborate? I tried to reference `Zcu.PerThread.refValue` but it doesn't call `singleConstPtrType`. I also retried replacing `many_item_ptr_type` with a call to `pt.singleConstPtrType(Type.fromInterned(array_ty))` but that trips some assertions so I'm not sure that I understood correctly.

Reply:
Sorry, the way GitHub showed this diff is really confusing -- I was commenting on the `orig_ty` line specifically.

`refValue` doesn't call `singleConstPtrType,` but it makes an equivalent type.

I think you can just replace this line with `.orig_ty = (try pt.singleConstPtrType(.fromInterned(array_ty)).toIntern(),`.

Reply:
Ah got it, thanks!